### PR TITLE
Faster array slicing

### PR DESCRIPTION
**Pull request by pitrou**
Despite the PR title, this also brings more generic optimizations to dependency computations, and various optimizations.

### CODE REVIEW

File: dask/core.py [hunk 1]
```diff
@@ -211,7 +211,7 @@ def get_dependencies(dsk, key=None, task=None, as_list=False):
     >>> get_dependencies(dsk, 'y')
     set(['x'])
 
-    >>> get_dependencies(dsk, 'z')  # doctest: +SKIP
+    >>> sorted(get_dependencies(dsk, 'z'))  # doctest: +SKIP
```
Comment (mrocklin) [inline dask/core.py]: Can we lose the `doctest: +SKIP`?

File: dask/array/slicing.py [hunk 2]
```diff
@@ -571,6 +585,42 @@ def insert_many(seq, where, val):
...
+                j += 1
+        right = "seq[%d:]" % j
+        left = "".join(left)
+        decl = decl.format(**locals())
+        ns = {}
+        exec(compile(decl, "<dynamic>", "exec"), ns, ns)
+        return ns['expand']
+
+
+def expander(where):
```
Comment (mrocklin) [inline dask/array/slicing.py]: Why separate this into two functions?

Comment (pitrou) [inline dask/array/slicing.py]: Because the `memoize` decorator needs hashable arguments (which lists are not).

File: dask/core.py (hunk 1 above)
Comment (pitrou) [inline dask/core.py]: I opted to remove the sorted() call instead.

File: dask/array/slicing.py (hunk 2 above)
Comment (mrocklin) [inline dask/array/slicing.py]: Try `@memoize(key=tuple)`

Also, should we be concerned about the memoized function holding on to too many inputs?

Comment (pitrou) [inline dask/array/slicing.py]: I don't think so, unless someone has a need for a large number of locations for `newaxis` in their indices.

Comment (pitrou) [inline dask/array/slicing.py]: > Try @memoize(key=tuple)

The signature for the key argument isn't very pretty (it takes a `args, kwargs` tuple). I wonder if it's not more readable to keep the two-argument variant.

**Comment by pitrou** [comment]
A bit embarassing: this PR makes https://nbviewer.jupyter.org/gist/shoyer/425cb3dc9101c235fe86 slower.

**Comment by pitrou** [comment]
I think I've fixed the regression. Also optimized `inline_functions` by using a recursive solution.

**Comment by pitrou** [comment]
I've now found out a way to make _deps() faster in an iterative way.

**Comment by mrocklin** [comment]
Why does creating the new list cause such a performance difference?  What differences are you seeing?

**Comment by pitrou** [comment]
The work list avoids both the recursion and the manual stack handling. The number of outer loop iterations is equal to the maximum task nesting level, which is usually very small.

**Comment by pitrou** [comment]
What difference I'm seeing: on
```
N = 1000
x = da.ones((N, N), chunks=(10, 10), name='x')
z = [x[i] for i in range(N)]
```
_deps() used to take 700 ms, it now only takes 500 ms.

**Comment by pitrou** [comment]
Sorry for the churn. I think this PR is now ready for review or merging.

File: dask/core.py [hunk 3]
```diff
@@ -224,29 +230,41 @@ def get_dependencies(dsk, key=None, task=None, as_list=False):
...
 
+def get_all_dependencies(dsk, tasks):
+    """ Get the immediate tasks on which the tasks depend.
+
+    Like get_dependencies(), but operating on multiple tasks at once,
+    and therefore significantly faster.
+    """
+    if not isinstance(tasks, list):
+        raise TypeError("Please provide a list of tasks")
+    return set(_deps(dsk, tasks))
```
Comment (mrocklin) [inline dask/core.py]: Given the new simplified structure of `get_dependencies` is this function still useful?

File: dask/core.py [hunk 4]
```diff
@@ -224,29 +230,41 @@ def get_dependencies(dsk, key=None, task=None, as_list=False):
...
+
+    Like get_dependencies(), but operating on multiple tasks at once,
+    and therefore significantly faster.
+    """
+    if not isinstance(keys, list):
+        raise TypeError("Please provide a list of keys")
+    if as_list:
+        return {k: _deps(dsk, dsk[k]) for k in keys}
+    else:
+        return {k: set(_deps(dsk, dsk[k])) for k in keys}
```
Comment (mrocklin) [inline dask/core.py]: Same question as above.  Do we still need this function?

**Comment by mrocklin** [comment]
@eriknw some of the performance optimizations to dask.core and dask.optimize might interest you.

File: dask/core.py (hunk 3 above)
Comment (pitrou) [inline dask/core.py]: Yes, because passing a bunch of tasks at once is more efficient than calling _deps() for each one.
(note _deps() can be very fast on trivial entries)

File: dask/core.py (hunk 4 above)
Comment (pitrou) [inline dask/core.py]: Ditto.

Comment (mrocklin) [inline dask/core.py]: In the case above I now see that we're sending a list of tasks.  In this case though we're just using a dict comprehension.  That dict-comprehension could presumably be inlined.  I suppose the gains here are based on the type checks and avoiding the extra function call?

Comment (pitrou) [inline dask/core.py]: Yes, they are. We could further inline the dict comprehension, but at the price of code duplication, so I'm not sure that it's a good idea.

**Comment by pitrou** [comment]
Hmm, I'm not sure why https://github.com/dask/dask/pull/1731/commits/831be05344fc2e0a7df8c95f41290450d36f4dd1 produced the following failure on AppVeyor:
https://ci.appveyor.com/project/dask-ci/dask/build/1.0.222#L490

Could it be an unrelated intermittent failure? One possible explanation is that `empty()` returned some NaNs or Infs and allclose() failed.

**Comment by mrocklin** [comment]
Yes, I suspect that this is caused by empty.  We should probably replace with random.random

File: dask/core.py (hunk 4 above)
Comment (mrocklin) [inline dask/core.py]: Ah, no, I was suggesting that instead we remove the `get_all_dependencies_per_key` function entirely and instead replace it with something like `{k: set(_deps(dsk, dsk[k])) for k in keys}`

Comment (pitrou) [inline dask/core.py]: Ah... I'll take a look.

Comment (pitrou) [inline dask/core.py]: `_deps()` is a private API, should we make it public? Use `get_dependencies` instead (at the price of a small slowdown - `cull()` seems to become 15% slower)?

Comment (mrocklin) [inline dask/core.py]: Thoughts on removing `_deps` entirely and folding it into `get_dependencies`?  We could place the burden of arg/task inputs and list/set outputs on the caller.

Comment (mrocklin) [inline dask/core.py]: This can also happen later.  Mostly I want to push on the principle of "there is a maintenance cost to having lots of functions that do similar things".  If we can have a few of these that compose well with standard language constructs like dictionary comprehensions then I think that dask.core will remain an easy place to get started for newish developers.

Comment (pitrou) [inline dask/core.py]: Definitely agreed. I'll try to fold `_deps` into `get_dependencies`.

**Comment by eriknw** [comment]
Thanks for the ping, I'll take a look when convenient.

**Comment by mrocklin** [comment]
Two thoughts:

1.  _deps is also used a couple of times within dask/distribued
2.  Thoughts on removing the `as_list=` keyword parameter and instead having the caller do this expilcitly as in `set(get_dependencies(dsk, task))`

**Comment by pitrou** [comment]
For distributed, see https://github.com/dask/distributed/pull/616.
`as_list` feels ok to me right now, but we can revisit it later. Returning a set by default makes more sense because the result is naturally unordered, and in most cases you don't care about duplicates.

**Comment by mrocklin** [comment]
OK

**Comment by mrocklin** [comment]
This seems fine to me.  I recommend that we wait a bit for https://github.com/dask/distributed/pull/616 to get in and in case @eriknw has some time for review.  Otherwise I plan to merge tomorrow morning (around 18 hours from now)

File: .travis.yml [hunk 5]
```diff
@@ -26,12 +26,14 @@ install:
...
-  - conda install -c conda-forge numpy=$NUMPY scipy pytables h5py bcolz pytest coverage toolz scikit-learn cytoolz chest blosc cython psutil ipython
-  - if [[ $PYTHON != '3.3' ]]; then conda install -c conda-forge pandas=$PANDAS distributed cloudpickle bokeh; fi
+  - conda install -c conda-forge numpy=$NUMPY scipy pytables h5py bcolz pytest coverage toolz scikit-learn cytoolz chest blosc cython psutil ipython python-lmdb
+  - if [[ $PYTHON != '3.3' ]]; then conda install -c conda-forge pandas=$PANDAS cloudpickle bokeh; fi
   - if [[ $PYTHON == '3.3' ]]; then pip install cloudpickle pandas==$PANDAS; fi
   - if [[ $PYTHON == '2.7' ]]; then pip install backports.lzma mock; fi
   - pip install git+https://github.com/mrocklin/partd --upgrade
   - pip install git+https://github.com/mrocklin/cachey --upgrade
+  - pip install git+https://github.com/dask/zict --upgrade
+  - pip install git+https://github.com/dask/distributed --upgrade
```
Comment (mrocklin) [inline .travis.yml]: Why are these necessary?

Comment (pitrou) [inline .travis.yml]: I'm trying to fix the following failure:
https://travis-ci.org/dask/dask/jobs/172618031#L731
I'm not sure why the build worked previously.

Comment (pitrou) [inline .travis.yml]: Now I'm getting another similar issue with `s3fs`: https://travis-ci.org/dask/dask/jobs/172623691#L751
Do you have an idea what may be happening?

Comment (mrocklin) [inline .travis.yml]: s3fs is included as a dependency in the `distributed` conda package but not in the PyPI package.  We keep the PyPI packages lightweight.

File: dask/core.py [hunk 6]
```diff
@@ -224,27 +186,34 @@ def get_dependencies(dsk, key=None, task=None, as_list=False):
...
+
+    while work:
+        new_work = []
+        for w in work:
+            typ = type(w)
+            if typ is tuple and w and callable(w[0]):  # istask(w)
+                new_work += w[1:]
+            elif typ is list:
+                new_work += w
+            elif typ is dict:
```
Comment (pitrou) [inline dask/core.py]: It's not added, it's part of the `_deps()` algorithm, rewritten iteratively, and folded here.

Comment (eriknw) [inline dask/core.py]: I see, thanks.  I'm confused why we want to do this with dicts.  This was added in 05ddd68e566daa1b3eb124002d4e3cf53ee407c3.
```python
In [20]: d = {'x': 1, 'y': {'a': 'x'}}

In [21]: dask.core.get_dependencies(d, 'y')
Out[21]: {'x'}

In [22]: dask.core.get(d, ['y'])
Out[22]: ({'a': 'x'},)

In [23]: dask.get(d, ['y'])
---------------------------------------------------------------------------
KeyError: 'y'
```

Comment (mrocklin) [inline dask/core.py]: @eriknw we do this for the distributed scheduler.  It ended up being useful, though at the moment I don't particularly recall the original use case.

```python
In [1]: from distributed import Client
In [2]: c = Client()
In [3]: d = {'x': 1, 'y': {'a': 'x'}}
In [4]: c.get(d, 'y')
Out[4]: {'a': 1}
```

Comment (mrocklin) [inline dask/core.py]: I don't recall the original motivation.  Thoughts on adding it to the spec and backporting to the single-machine scheduler?

Comment (jcrist) [inline dask/core.py]: I'm slightly against it - it makes the spec more complicated, and may result in increased overhead (although that may be negligible). For kwargs that contain dask arguments (rare), we decompose a dict into tasks: `(dict, ['a'], ['x'])`, and everything works out as is.

Comment (mrocklin) [inline dask/core.py]: Supporting dict values seems like a natural extension to me.  More natural than the `(dict, ['a'], ['x'])` solution.

Comment (pitrou) [inline dask/core.py]: I agree the `dict` solution doesn't look pretty.

Comment (eriknw) [inline dask/core.py]: Here's a simple, non-recursive helper function that makes the intent explicit and doesn't require changing the spec:
```python
In [35]: def subdask(d):
    ...:     return (dict, (zip, tuple(d.keys()), list(d.values())))
    ...: 
    ...: d = {
    ...:     'a': 1,
    ...:     'b': 2,
    ...:     'c': subdask({'a': 'a', 'b': 'b'}),
    ...: }
    ...: 

In [36]: d
Out[36]: {'a': 1, 'b': 2, 'c': (dict, (<function zip>, ('a', 'b'), ['a', 'b']))}

In [37]: dask.core.get(d, 'c')
Out[37]: {'a': 1, 'b': 2}
```

Comment (pitrou) [inline dask/core.py]: By the way, I'm assuming this discussion isn't an issue for the current PR, right? :-)

Comment (eriknw) [inline dask/core.py]: Yeah.  The PR lgtm.  It keeps the existing behavior with dicts, which we can revisit elsewhere.

Comment (mrocklin) [inline dask/core.py]: https://github.com/dask/dask/pull/1748

**Comment by mrocklin** [comment]
Merging in 12 hours if no comment
